### PR TITLE
[UTOPIA-1288] only allow one idir user to review one program area selected role

### PR DESCRIPTION
**Pull request by binggaofw**
# Description

This PR includes the following proposed change(s):

- see title, if a user(based on IDIR) already review a program area role, then all other role will become read only for this user

Screenshot 2023-07-13 at 8 19 36 AM

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Updating Testing Framework(s)
- [ ] Version change

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Development Dependency Working Agreement
- [x] My code DOES NOT include the importing of new dependencies into the DPIA ecosystem
- [ ] My code DOES include the importing of new dependencies into the DPIA ecosystem
**If new dependencies are being introduced to the DPIA ecosystem:**
- [ ] The functionality of the dependency drastically reduces code complexity and makes my changes more easily maintainable and readible 
- [ ] The dependency being introduced does not contain multiple layers of nested dependencies introducing maintainability complexity to the DPIA ecosystem

## Frontend Development Changes
- [ ] N/A
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to project documentation or diagrams that reflect my changes
- [ ] New and existing unit tests pass locally with my changes
- [ ] My code follows Airbnb React Style Guidelines

## API Development Changes
- [x] N/A
- [ ] I have performed a self-review of my own code
- [ ] My code follows standards and practices outlined in the BC Government API Development Guidelines
- [ ] New and existing unit tests pass locally with my changes
- [ ] My changes includes Swagger documentation updates that reflect the changes I am introducing

## Definition of Done

Definition of Done

### CODE REVIEW

File: src/frontend/src/components/public/PIAFormTabs/review/index.tsx [hunk 1]
```diff
@@ -111,7 +110,20 @@ const PIAReview = ({ printPreview }: IReviewProps) => {
       true,
     );
   };
+  const allowUserReviewProgramArea = () => {
+    const userGuid = getGUID();
+    const selectedRoles = reviewForm?.programArea?.selectedRoles;
```
Comment (kushal-arora-fw) [inline src/frontend/src/components/public/PIAFormTabs/review/index.tsx:115]: please add a null check after this. If selectedRoles object does not exist return true.

Comment (binggaofw) [reply src/frontend/src/components/public/PIAFormTabs/review/index.tsx:115]: added the check

**Review by Bottle7 — APPROVED** [review]

**Comment by kushal-arora-fw** [comment]
The UX doesn't look good OR I guess, it was never defined for this scenario. 

If a user has filled in review for one of the program area, then other reviews look N/A,  blank and weird view state. I mean functionality works, and so approving.

**Review by kushal-arora-fw — APPROVED** [review]

**Comment by kushal-arora-fw** [comment]
But I would recommend having UX give it a look once.

File: src/frontend/src/components/public/PIAFormTabs/review/index.tsx [hunk 2]
```diff
@@ -246,12 +260,18 @@ const PIAReview = ({ printPreview }: IReviewProps) => {
                             className="d-flex align-items-center"
                             key={index}
                           >
-                            {Object(pia?.review?.programArea)?.reviews?.[role]
+                            {!allowUserReviewProgramArea() ||
```
Comment (antsand) [inline src/frontend/src/components/public/PIAFormTabs/review/index.tsx:263]: this logic is not very clear. Even if one of the program area reviewbyguid == to the current user guid it returns  false and they see the view mode? Seems convoluted.

File: src/frontend/src/components/public/PIAFormTabs/review/index.tsx [hunk 3]
```diff
@@ -113,7 +112,22 @@ const PIAReview = ({ printPreview }: IReviewProps) => {
       true,
     );
   };
+  const allowUserReviewProgramArea = () => {
+    const userGuid = getGUID();
+    const selectedRoles = reviewForm?.programArea?.selectedRoles;
+    // if selectedRoles is null means none of selectedRole got reviewed so return true
+    // otherwise loop all the role in reviews part to see if the current user already did review
+    if (selectedRoles === null) return true;
+    for (const role of selectedRoles) {
```
Comment (antsand) [inline src/frontend/src/components/public/PIAFormTabs/review/index.tsx:121]: should this be done on a case by case bases? this function is called to loop and if even one matches progamareaguiluserguid the whole review page turns to edit mode? Not very clear

**Review by antsand — APPROVED** [review]

**Pull request merged**